### PR TITLE
Fix a lot of stuff on user profile

### PR DESCRIPTION
**Pull request by TheFehr**

### CODE REVIEW

**Comment by TheFehr** [comment]
Closes #91

File: api/app/models/service.rb [hunk 1]
```diff
@@ -86,4 +87,10 @@ def beginning_is_monday
   def ending_is_friday
     errors.add(:ending, :not_a_friday) unless ending.present? && ending.wday == FRIDAY_WEEKDAY
   end
+
+  def length_is_valid
+    return if ending.blank? || beginning.blank? || last_civil_service?
+
+    errors.add(:service_days, :invalid_length) if (ending - beginning).to_i < 25
```
Comment (lukasbischof) [inline api/app/models/service.rb]: Please move 25 to constant

File: api/spec/models/service_spec.rb [hunk 2]
```diff
@@ -169,7 +169,10 @@
   end
 
   describe 'beginning_is_monday validation' do
-    subject { build(:service, beginning: beginning).tap(&:validate).errors.added? :beginning, :not_a_monday }
+    subject do
+      build(:service, beginning: beginning, ending: (beginning + 4.weeks).at_end_of_week - 2.days)
```
Comment (lukasbischof) [inline api/spec/models/service_spec.rb]: I‘d prefer ending as a let in favour of multiline let

File: api/spec/models/service_spec.rb [hunk 3]
```diff
@@ -207,4 +210,40 @@
...
+    end
+
+    context 'when service is last' do
+      let(:service) { build(:service, beginning: beginning, ending: ending, user: user, service_type: :last) }
+
+      context 'when service has a valid length' do
+        it { is_expected.to be false }
+      end
+
+      context 'when service has a invalid length' do
```
Comment (lukasbischof) [inline api/spec/models/service_spec.rb]: Please use something more descriptive like ‚when service length is below 4 weeks‘

File: api/spec/requests/v1/services_controller_spec.rb [hunk 4]
```diff
@@ -138,7 +137,7 @@
                 identification_number: Service.last.identification_number,
                 name: Service.last.service_specification.name
               }
-            )
+            ).transform_values { |value| value.is_a?(Symbol) ? value.to_s : value }
```
Comment (lukasbischof) [inline api/spec/requests/v1/services_controller_spec.rb]: You could also use something like

```suggestion
            ).transform_values { |value| value.try(:to_s) || value }
```

File: frontend/src/form/FormView.tsx [hunk 5]
```diff
@@ -20,25 +20,24 @@ interface Props<T> extends FormViewProps<T> {
...
-            <FormikSubmitDetector {...formikProps}>
-              <Prompt when={!this.props.submitted && formikProps.dirty} message={() => 'Änderungen verwerfen?'} />
-              {this.props.render(formikProps)}
-            </FormikSubmitDetector>
-          )}
-        />
-        {children}
+    return (
+      <IziviContent card={this.props.card} loading={loading} title={title}>
+        {this.props.loading ? null :
```
Comment (lukasbischof) [inline frontend/src/form/FormView.tsx]:
```suggestion
        {!this.props.loading &&
```

File: frontend/src/stores/serviceStore.ts [hunk 6]
```diff
@@ -1,5 +1,7 @@
 import { action, computed, observable } from 'mobx';
+import moment from 'moment';
 import { Service, ServiceCollection } from '../types';
+import {apiDateFormat} from './apiStore';
```
Comment (lukasbischof) [inline frontend/src/stores/serviceStore.ts]: please space

File: frontend/src/stores/serviceStore.ts [hunk 7]
```diff
@@ -84,12 +85,18 @@ export class ServiceStore extends DomainStore<Service, ServiceCollection> {
     const response = await this.mainStore.api.put<ServiceCollection[]>('/services/' + service.id, service);
     this.services = response.data;
   }
+
+  private async callServiceCalculator<ReturnType>(url: string) {
+    const response = await this.mainStore.api.
+    get<ReturnType>(url);
```
Comment (lukasbischof) [inline frontend/src/stores/serviceStore.ts]: Why is this on a new line?
Anyway, I would preder this method in serviceStore as static method and URLs as helpers

File: frontend/src/types.ts [hunk 8]
```diff
@@ -1,4 +1,4 @@
-import { FormikBag } from 'formik';
+import {FormikBag} from 'formik';
```
Comment (lukasbischof) [inline frontend/src/types.ts]: Why?

File: frontend/src/types.ts [hunk 9]
```diff
@@ -33,34 +33,33 @@ export interface PaymentExpenseSheet {
...
-  id?: number;
+  id: number;
   bank_account_number: string;
-  beginning: string;
+  beginning: Date;
   clothing_expenses: number;
-  clothing_expenses_comment: string;
-  company_holiday_comment: string;
+  clothing_expenses_comment: null;
+  company_holiday_comment: null;
```
Comment (lukasbischof) [inline frontend/src/types.ts]: Why is this always null?

File: frontend/src/types.ts [hunk 10]
```diff
@@ -195,9 +195,9 @@ export interface ServiceCollection {
   ending: string | null;
   confirmation_date: string | null;
   service_specification: {
-    identification_number: string;
-    name: string | null;
-    short_name: string | null;
+    identification_number: number;
```
Comment (lukasbischof) [inline frontend/src/types.ts]: Is it really a number?

File: frontend/src/views/expense_sheets/ExpenseSheetUpdate.tsx [hunk 11]
```diff
@@ -3,6 +3,7 @@ import { inject, observer } from 'mobx-react';
 import * as React from 'react';
 import { RouteComponentProps } from 'react-router';
 import { ExpenseSheetStore } from '../../stores/expenseSheetStore';
+import {ServiceSpecificationStore} from '../../stores/serviceSpecificationStore';
```
Comment (lukasbischof) [inline frontend/src/views/expense_sheets/ExpenseSheetUpdate.tsx]: space

File: frontend/src/views/expense_sheets/expense_sheet_form/ExpenseSheetForm.tsx [hunk 12]
```diff
@@ -6,7 +6,8 @@ import Form from 'reactstrap/lib/Form';
 import { FormView, FormViewProps } from '../../../form/FormView';
 import { ExpenseSheetStore } from '../../../stores/expenseSheetStore';
 import { MainStore } from '../../../stores/mainStore';
-import { ExpenseSheet, ExpenseSheetHints, FormValues, Service } from '../../../types';
+import {ServiceSpecificationStore} from '../../../stores/serviceSpecificationStore';
+import {ExpenseSheet, ExpenseSheetHints, FormValues, Service, ServiceSpecification} from '../../../types';
```
Comment (lukasbischof) [inline frontend/src/views/expense_sheets/expense_sheet_form/ExpenseSheetForm.tsx]: space

File: frontend/src/views/expense_sheets/expense_sheet_form/ExpenseSheetForm.tsx [hunk 13]
```diff
@@ -19,32 +20,44 @@ type Props = {
...
-    return (
+    if (!loading) {
+      serviceSpecificationStore!.fetchOne(service.service_specification_id).then(() => {
+        this.setState({
+          loading: false,
+        });
+      });
+    }
+
+    return empty(expenseSheet) || this.props.loading ?
```
Comment (lukasbischof) [inline frontend/src/views/expense_sheets/expense_sheet_form/ExpenseSheetForm.tsx]: Can you extract this in a normal if/else. Is a bit long for ternary

File: frontend/src/views/expense_sheets/expense_sheet_form/ExpenseSheetFormHeader.tsx [hunk 14]
```diff
@@ -1,6 +1,6 @@
 import * as React from 'react';
 import Badge from 'reactstrap/lib/Badge';
-import { ExpenseSheetState, Service } from '../../../types';
```
Comment (lukasbischof) [inline frontend/src/views/expense_sheets/expense_sheet_form/ExpenseSheetFormHeader.tsx]: ...

File: frontend/src/views/users/service_subform/ServiceOverviewTable.tsx [hunk 15]
```diff
@@ -1,17 +1,18 @@
-import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
 import moment from 'moment';
 import * as React from 'react';
-import { WithSheet } from 'react-jss';
-import { Link } from 'react-router-dom';
```
Comment (lukasbischof) [inline frontend/src/views/users/service_subform/ServiceOverviewTable.tsx]: ...

File: frontend/src/views/users/service_subform/ServiceSubform.tsx [hunk 16]
```diff
@@ -2,6 +2,7 @@ import { inject } from 'mobx-react';
 import * as React from 'react';
 import injectSheet, { WithSheet } from 'react-jss';
 import Button from 'reactstrap/lib/Button';
+import {ExpenseSheetStore} from '../../../stores/expenseSheetStore';
```
Comment (lukasbischof) [inline frontend/src/views/users/service_subform/ServiceSubform.tsx]: ...

File: api/app/controllers/v1/services_controller.rb [hunk 17]
```diff
@@ -6,11 +6,10 @@ class ServicesController < ApplicationController
     include V1::Concerns::ParamsAuthenticatable
     include V1::Concerns::JsonAndPdfRespondable
 
-    PERMITTED_SERVICE_SPECIFICATION_PARAMS = %i[service_specification_identification_number].freeze
```
Comment (lukasbischof) [inline api/app/controllers/v1/services_controller.rb]: Why did you remove this?

File: api/app/controllers/v1/services_controller.rb [hunk 18]
```diff
@@ -94,20 +93,7 @@ def sanitize_service_params(permitted_params)
       permitted_params[:user_id] = current_user.id unless current_user.admin?
       permitted_params.except!(:confirmation_date) unless current_user.admin?
 
-      associate_service_specification(permitted_params)
-    end
-
-    def associate_service_specification(service_params)
```
Comment (lukasbischof) [inline api/app/controllers/v1/services_controller.rb]: I thought that the idea was that we do not expose the db id to frontend as the identification number is already a unique index

File: api/spec/requests/v1/services_controller_spec.rb (hunk 4 above)
Comment (TheFehr) [reply api/spec/requests/v1/services_controller_spec.rb]: Can't do because then numbers would also get converted, which is not what this code should do.

File: frontend/src/form/FormView.tsx [hunk 19]
```diff
@@ -19,10 +19,10 @@ interface Props<T> extends FormViewProps<T> {
 
 export class FormView<Values, ExtraProps = {}> extends React.Component<FormikConfig<Values> & ExtraProps & Props<Values>> {
   render() {
-    const { loading, title, children, ...rest } = this.props as any;
+    const { loading, title, children, ...rest } = this.props;
     return (
       <IziviContent card={this.props.card} loading={loading} title={title}>
-        {this.props.loading &&
+        {!this.props.loading &&
```
Comment (lukasbischof) [inline frontend/src/form/FormView.tsx]: 🥺
